// balena-settings-client/build/defaults.js
"use strict";
/*
Copyright 2016-17 Balena

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
var hidepath = require("hidepath");
var userHome = require("home-or-tmp");
var path_1 = require("path");
module.exports = {
    /**
     * @property {String} balenaUrl - Balena API url
     * @memberof defaults
     */
    balenaUrl: 'balena-cloud.com',
    /**
     * @property {Function} apiUrl - Balena API url
     * @memberof defaults
     */
    apiUrl: function () {
        return "https://api.".concat(this.balenaUrl);
    },
    /**
     * @property {Function} vpnUrl - Balena VPN url
     * @memberof defaults
     */
    vpnUrl: function () {
        return "vpn.".concat(this.balenaUrl);
    },
    /**
     * @property {Function} registryUrl - Balena Registry url
     * @memberof defaults
     */
    registryUrl: function () {
        return "registry.".concat(this.balenaUrl);
    },
    /**
     * @property {Function} registry2Url - Balena Registry 2 url
     * @memberof defaults
     */
    registry2Url: function () {
        return "registry2.".concat(this.balenaUrl);
    },
    /**
     * @property {Function} deltaUrl - Balena Delta url
     * @memberof defaults
     */
    deltaUrl: function () {
        return "https://delta.".concat(this.balenaUrl);
    },
    /**
     * @property {Function} dashboardUrl - Balena dashboard url
     * @memberof defaults
     */
    dashboardUrl: function () {
        return "https://dashboard.".concat(this.balenaUrl);
    },
    /**
     * @property {Function} proxyUrl - Balena Proxy url
     * @memberof defaults
     */
    proxyUrl: function () {
        if (this.balenaUrl === 'balena-cloud.com') {
            return 'balena-devices.com';
        }
        else if (this.balenaUrl === 'balena-staging.com') {
            return 'balena-staging-devices.com';
        }
        else if (this.balenaUrl === 'resin.io') {
            return 'resindevice.io';
        }
        return "devices.".concat(this.balenaUrl);
    },
    /**
     * @property {Function} tunnelUrl - Balena Tunnel url
     * @memberof defaults
     */
    tunnelUrl: function () {
        return "tunnel.".concat(this.balenaUrl);
    },
    /**
     * @property {String} dataDirectory - data directory path
     * @memberof defaults
     */
    dataDirectory: (0, path_1.join)(userHome, hidepath('balena')),
    /**
     * @property {String} projectsDirectory - projects directory path
     * @memberof defaults
     */
    projectsDirectory: (0, path_1.join)(userHome, 'BalenaProjects'),
    /**
     * @property {Function} cacheDirectory - cache directory path
     * @memberof defaults
     */
    cacheDirectory: function () {
        return (0, path_1.join)(this.dataDirectory, 'cache');
    },
    /**
     * @property {Function} binDirectory - binary directory path
     * @memberof defaults
     */
    binDirectory: function () {
        return (0, path_1.join)(this.dataDirectory, 'bin');
    },
    /**
     * @property {Number} imageCacheTime - image cache time
     * @memberof defaults
     */
    imageCacheTime: 1 * 1000 * 60 * 60 * 24 * 7,
    /**
     * @property {Number} tokenRefreshInterval - token refresh interval
     * @memberof defaults
     */
    tokenRefreshInterval: 1 * 1000 * 60 * 60,
    /**
     * @property {String} apiKeyVariable - api key environment variable
     * @memberof defaults
     */
    apiKeyVariable: 'BALENA_API_KEY',
};
