// @balena/sbvr-types/out/types/sha.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.compare = exports.validate = exports.validateSync = exports.types = void 0;
const TypeUtils = require("../type-utils");
let sha256;
try {
    const crypto = require('crypto');
    sha256 = (value) => {
        const hash = crypto.createHash('sha256');
        hash.update(value);
        return `$sha256$${hash.digest('base64')}`;
    };
}
catch {
    const shajs = require('sha.js');
    sha256 = (value) => {
        const hash = shajs('sha256');
        hash.update(value);
        return `$sha256$${hash.digest('base64')}`;
    };
}
exports.types = {
    postgres: 'CHAR(54)',
    mysql: 'CHAR(54)',
    websql: 'CHAR(54)',
    odata: {
        name: 'Edm.String',
    },
};
exports.validateSync = sha256;
exports.validate = TypeUtils.validate.checkRequired((value) => {
    if (typeof value !== 'string') {
        throw new Error('is not a string');
    }
    return sha256(value);
});
const compare = async (value, result) => {
    return sha256(value) === result;
};
exports.compare = compare;
